tests/SerializerTest.php

```diff
@@ -0,0 +1,178 @@
+<?php
+/**
+ * Copyright 2016 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\GAX\UnitTests;
+
+use Google\Api\HttpRule;
+use Google\GAX\Serializer;
+use Google\Protobuf\Any;
+use Google\Protobuf\Field;
+use Google\Protobuf\FieldMask;
+use Google\Protobuf\ListValue;
+use Google\Protobuf\Struct;
+use Google\Protobuf\Value;
+use Google\Rpc\Status;
+
+/**
+ * @group core
+ */
+class SerializerTest extends \PHPUnit_Framework_TestCase
+{
+    private function backAndForth($message, $arrayStructure)
+    {
+        $serializer = new Serializer();
+        $klass = get_class($message);
+
+        $serializedMessage = $serializer->encodeMessage($message);
+        $deserializedMessage = $serializer->decodeMessage(new $klass(), $serializedMessage);
+        $this->assertEquals($arrayStructure, $serializedMessage);
+        $this->assertEquals($message, $deserializedMessage);
+
+        $deserializedStructure = $serializer->decodeMessage(new $klass(), $arrayStructure);
+        $reserializedStructure = $serializer->encodeMessage($deserializedStructure);
+        $this->assertEquals($message, $deserializedStructure);
+        $this->assertEquals($arrayStructure, $reserializedStructure);
+    }
+
+    public function testStatusMessage()
+    {
+        $details = [new Any()];
+        $message = new Status();
+        $message->setMessage("message");
+        $message->setCode(0);
+        $message->setDetails($details);
+        $this->backAndForth($message, [
+            'message' => 'message',
+            'code' => 0,
+            'details' => [
+                [
+                    'typeUrl' => '',
+                    'value' => '',
+                ],
+            ]
+        ]);
+    }
+
+    public function testHttpRule()
+    {
+        $message = new HttpRule();
+        $this->backAndForth($message, [
+            'selector' => '',
+            'body' => '',
+            'additionalBindings' => [],
+        ]);
```

Comment:
The array is what we expect `$message` to encode to, and vice versa. I can make this more explicit